### PR TITLE
[ETD-160] Added packetizer for cd

### PR DESCRIPTION
**Pull request by Ficzer**

### CODE REVIEW

File: src/Broker/Broker.c [hunk 1]
```diff
@@ -80,6 +81,10 @@ Broker_initialize_packetizers_functions()
     packetizers_functions[PACKETIZER_THIN].init = &ThinPacketizer_init;
     packetizers_functions[PACKETIZER_THIN].packetize = &ThinPacketizer_packetize;
     packetizers_functions[PACKETIZER_THIN].depacketize = &ThinPacketizer_depacketize;
+    packetizers_functions[PACKETIZER_DEVICE_PROVIDED].headerSize = THIN_SPACE_PACKET_PRIMARY_HEADER_SIZE;
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:84]: Will this be the same for all device provided packetizers?

File: src/Packetizer/DeviceProvidedPacketizer.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "DeviceProvidedPacketizer.h"
+
+void
+DeviceProvidedPacketizer_init(Packetizer* const self, const enum SystemBus busId)
+{
+    packetizer_init_function packetizer_init = getDeviceProvidedPacketizerInitFunction(busId);
```
Comment (Lurkerpas) [inline src/Packetizer/DeviceProvidedPacketizer.c:28]: This can return NULL. This should not happen, so either assert, or check and early return.

File: src/RuntimeMocks/device_provided_packetizer.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#include <system_spec.h>
+#include <Packetizer.h>
+
+extern packetizer_init_function bus_to_device_provided_packetizer_init_function[SYSTEM_BUSES_NUMBER];
```
Comment (Lurkerpas) [inline src/RuntimeMocks/device_provided_packetizer.h:4]: I don't think these externs are needed.

File: src/Packetizer/DeviceProvidedPacketizer.c (hunk 2 above)
Comment (Ficzer) [reply src/Packetizer/DeviceProvidedPacketizer.c:28]: Added NULL check

File: src/RuntimeMocks/device_provided_packetizer.h (hunk 3 above)
Comment (Ficzer) [reply src/RuntimeMocks/device_provided_packetizer.h]: Removed unnecessary mock parts.

File: src/Broker/Broker.c (hunk 1 above)
Comment (Ficzer) [reply src/Broker/Broker.c:84]: Done

File: src/Packetizer/CCSDSPacketizer.c [hunk 4]
```diff
@@ -31,10 +31,11 @@
 #include "SpacePacketInternal.h"
 
 void
-CCSDS_Packetizer_init(Packetizer* const self, const enum SystemBus busId)
+CCSDS_Packetizer_init(Packetizer* const self, const enum SystemBus busId, size_t* const headerSize)
 {
     // Unused in this implementation
     (void)busId;
+    (void)headerSize;
```
Comment (Lurkerpas) [inline src/Packetizer/CCSDSPacketizer.c:38]: It would be nice if a correct value was returned.

File: src/Packetizer/Packetizer.c [hunk 5]
```diff
@@ -30,10 +30,11 @@
 #include "SpacePacketInternal.h"
 
 void
-Packetizer_init(Packetizer* const self, const enum SystemBus busId)
+Packetizer_init(Packetizer* const self, const enum SystemBus busId, size_t* const headerSize)
 {
     // Unused in this implementation
     (void)busId;
+    (void)headerSize;
```
Comment (Lurkerpas) [inline src/Packetizer/Packetizer.c:37]: It would be nice if a correct value was returned.

File: src/Broker/Broker.c [hunk 6]
```diff
@@ -107,7 +108,11 @@ Broker_initialize(enum SystemBus valid_buses[SYSTEM_BUSES_NUMBER])
             // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85398
             enum PacketizerCfg packetizer_type = bus_to_packetizer_cfg[bus_id];
             packetizer_init_function packetizer_init = packetizers_functions[packetizer_type].init;
-            packetizer_init(&packetizers_data[bus_id], bus_id);
+            packetizer_init(&packetizers_data[bus_id], bus_id, &headerSize);
+
+            if(packetizer_type == PACKETIZER_DEVICE_PROVIDED) {
+                packetizers_functions[packetizer_type].headerSize = headerSize;
```
Comment (Lurkerpas) [inline src/Broker/Broker.c:114]: This is a simpler solution than the one we discussed - does this one handle a situation in which there are two devices, with different packetizers? Won't the last one overwrite the size reported by the previous ones?

File: src/Broker/Broker.c (hunk 1 above)
Comment (Ficzer) [reply src/Broker/Broker.c:84]: THIN_SPACE_PACKET_PRIMARY_HEADER_SIZE set as default, It is changed later in DeviceProvidedPacketizer

File: src/Packetizer/CCSDSPacketizer.c (hunk 4 above)
Comment (Ficzer) [reply src/Packetizer/CCSDSPacketizer.c]: Done

File: src/Packetizer/ThinPacketizer.c [hunk 7]
```diff
@@ -34,11 +34,12 @@ void thinWritePacketDataLength(uint8_t* const packetPointer, const size_t dataSi
 size_t thinReadPacketDataLength(const uint8_t* const packetPointer);
 
 void
-ThinPacketizer_init(Packetizer* const self, const enum SystemBus busId)
+ThinPacketizer_init(Packetizer* const self, const enum SystemBus busId, size_t* const headerSize)
 {
     // Unused in this implementation
     (void)self;
     (void)busId;
+    (void)headerSize;
```
[the remark this answers is not in the document]
Comment (Ficzer) [reply src/Packetizer/ThinPacketizer.c]: Done

File: src/Broker/Broker.c (hunk 6 above)
Comment (Ficzer) [reply src/Broker/Broker.c]: Added extra array with CD packetizer header, the header size is get by busId in DeviceProvidedPacketizer

File: src/Packetizer/DeviceProvidedPacketizer.c [hunk 8]
```diff
@@ -40,13 +42,20 @@ DeviceProvidedPacketizer_packetize(Packetizer* const self,
...
+    if(headerSize <= 0) {
+        return 0;
+    }
+
     if(packetizer_packetize == NULL) {
         return 0;
     }
 
-    return packetizer_packetize(self, packetType, busId, source, destination, packetPointer, dataOffset, dataSize);
+    return packetizer_packetize(self, packetType, busId, source, destination, packetPointer, headerSize, dataSize);
```
Comment (Lurkerpas) [inline src/Packetizer/DeviceProvidedPacketizer.c:58]: Why this change?

Comment (Ficzer) [reply src/Packetizer/DeviceProvidedPacketizer.c]: Removed

File: src/Packetizer/DeviceProvidedPacketizer.c [hunk 9]
```diff
@@ -40,13 +42,20 @@ DeviceProvidedPacketizer_packetize(Packetizer* const self,
                                    const size_t dataOffset,
                                    const size_t dataSize)
 {
+    (void)dataOffset;
+
+    size_t headerSize = getDeviceProvidedPacketizerHeaderSize(busId);
     packetizer_packetize_function packetizer_packetize = getDeviceProvidedPacketizerPacketizeFunction(busId);
 
+    if(headerSize <= 0) {
```
[the remark this answers is not in the document]
Comment (Ficzer) [reply src/Packetizer/DeviceProvidedPacketizer.c]: Done

File: src/Broker/Broker.c (hunk 1 above)
Comment (Ficzer) [reply src/Broker/Broker.c]: Set to 0

File: src/Packetizer/DeviceProvidedPacketizer.h [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                                          const uint8_t* const packetPointer,
+                                          const size_t packetSize,
+                                          const enum SystemBus busId,
+                                          uint16_t* const source,
+                                          uint16_t* const destination,
+                                          size_t* const dataOffset,
+                                          size_t* const dataSize,
+                                          int32_t* const errorCode);
+
+#endif // THIN_PACKETIZER_H
```
Comment (psurm) [inline src/Packetizer/DeviceProvidedPacketizer.h:97]: DEVICE_PROVIDED_PACKETIZER_H

Comment (Ficzer) [reply src/Packetizer/DeviceProvidedPacketizer.h]: Done

**Review by psurm — APPROVED** [review]
